agent/consul/grpc_client.go

```diff
@@ -66,11 +55,47 @@ func (c *GRPCClient) GRPCConn(server *metadata.Server) (*grpc.ClientConn, error)
 	c.grpcConnsLock.Lock()
 	defer c.grpcConnsLock.Unlock()
 
-	co, err := grpc.Dial(addr, grpc.WithInsecure(), grpc.WithDialer(dialGRPC), grpc.WithDisableRetry())
+	dialer := newDialer(server.UseTLS, server.Datacenter, c.tlsWrap)
+	co, err := grpc.Dial(addr, grpc.WithInsecure(), grpc.WithDialer(dialer), grpc.WithDisableRetry())
 	if err != nil {
 		return nil, err
 	}
 
 	c.grpcConns[addr] = conn
 	return co, nil
 }
+
+// newDialer returns a gRPC dialer function that conditionally wraps the connection
+// with TLS depending on the given useTLS value.
+func newDialer(useTLS bool, datacenter string, wrapper tlsutil.DCWrapper) func(string, time.Duration) (net.Conn, error) {
+	return func(addr string, _ time.Duration) (net.Conn, error) {
+		conn, err := net.Dial("tcp", addr)
+		if err != nil {
+			return nil, err
+		}
+
+		// Check if TLS is enabled
+		if useTLS && wrapper != nil {
```

Comment:
If `useTLS` is true but `wrapper` is nil this will silently degrade to plaintext which seems like CVE material. Should we explicitly error in that case instead?

Reply:
That's probably a good call - this is pulled from the existing rpc pool stuff for wrapping a connection in TLS so we should probably error there too: https://github.com/hashicorp/consul/blob/master/agent/pool/pool.go#L308

Reply:
Thats a good point. this is giving me some headaches - `useTLS` shouldn't be necessary, it should be handled by the `tlsConfigurator` instead which would make it much easier to reason about. I know we have that flag in a couple of places and it complicates matters. I created an issue for it: https://github.com/hashicorp/consul/issues/6337.

I think it might be possible that useTLS is set to true on startup and due to TLS config reloading the wrapper returns nil.

Reply:
I thought about the CVE potential a bit more. I think the degration is only possible if the configuration starts out with TLS setup, then the TLS bits are removed from the config and the client reloads the TLS config. This is the only way I can think of, how this could happen. I don't think it would be a CVE since the user took the TLS out of the configuration and reloaded. I would say going unencrypted shouldn't be a surprise after that.

Reply:
Thanks for thinking it through @i0rek. 

When I said "CVE Material" I was not so much thinking it's already a vuln and more that it's a bit of logic that makes some assumptions about other code elsewhere (that wrapper is never nil when useTLS is true) that could be brittle to changes in other parts of the codebase.

I'd vote for switching only on `useTLS` and explicitly handling the case wrapper is nil. That should never happen now but would be better if that became an error rather than a CVE later after some refactor.

Reply:
@i0rek could you take a look again and make sure the TLS updates look sane? I think I made all the necessary changes to support reloading.